src/indexer.c

```diff
@@ -113,6 +113,32 @@ void *ofi_idx_remove(struct indexer *idx, int index)
 	return item;
 }
 
+void *ofi_idx_remove_ordered(struct indexer *idx, int index)
+{
+	struct ofi_idx_entry *entry;
+	void *item;
+	int indx;
+	int entry_index = ofi_idx_entry_index(index);
+
+	entry = idx->array[ofi_idx_array_index(index)];
+	item = entry[entry_index].item;
+	assert(item != NULL);
+	entry[entry_index].item = NULL;	
+	if (index < idx->free_list || entry[entry_index].next == idx->free_list) {
```

Comment:
entry is in use, so it's next pointer isn't valid.  The second part of this if check isn't needed.  Maybe you're wanting to check if the free_list is empty?

Reply:
The last location of the 'entry[]' always seems to point to index  0 ( from the grow function) to which the idx->free_list also points to when all entries in (idx->array[i]) are filled up. At this point if the last entry is requested to be removed then the index > idx->free_list, but entry[index].next ==idx->free_list. This was the case i was trying to handle here.

Reply:
0 is equivalent to null at the end of the free list.  I think this check is what you want:
```
if (idx->free_list == 0 || index < idx->free_list)
```
This should insert entry at the head of the free list if it is empty or if the entry's index comes first.

If it makes things clearer, we could define free_list_empty() and move the check in there.